src/custom/pages/Swap/components/ApproveButton.tsx

```diff
@@ -0,0 +1,224 @@
+import { useERC20PermitFromTrade } from 'hooks/useERC20Permit'
+import TradeGp from 'state/swap/TradeGp'
+import { Currency, Percent } from '@uniswap/sdk-core'
+import { BigNumber } from '@ethersproject/bignumber'
+import { SwapConfirmState } from 'pages/Swap/state/swapConfirmAtom'
+import { ThemeContext } from 'styled-components/macro'
+import usePrevious from '@src/hooks/usePrevious'
+
+export interface ApproveButtonProps {
+  currencyIn: Currency | undefined | null
+  trade: TradeGp | undefined
+  allowedSlippage: Percent
+  transactionDeadline: BigNumber | undefined
+  setSwapState: (state: SwapConfirmState) => void
+  swapConfirmState: SwapConfirmState
+  SwapButton: React.FC<SwapButtonProps>
+  isExpertMode: boolean
+  handleSwap: () => void
+  isValid: boolean
+  isGettingNewQuote: boolean
+  swapBlankState: boolean
+  approvalState: ApprovalState
+  approveCallback: (params?: OptionalApproveCallbackParams) => Promise<void>
+  approvalSubmitted: boolean
+  setApprovalSubmitted: (state: boolean) => void
+}
+
+export function ApproveButton(props: ApproveButtonProps) {
+  const {
+    trade,
+    allowedSlippage,
+    transactionDeadline,
+    setSwapState,
+    swapConfirmState,
+    currencyIn,
+    SwapButton,
+    isExpertMode,
+    handleSwap,
+    isValid,
+    isGettingNewQuote,
+    swapBlankState,
+    approvalState,
+    approveCallback,
+    approvalSubmitted,
+    setApprovalSubmitted,
+  } = props
+
+  const theme = useContext(ThemeContext)
+
+  const { showConfirm, tradeToConfirm } = swapConfirmState
+
+  const {
+    state: signatureState,
+    // signatureData,
+    gatherPermitSignature,
+  } = useERC20PermitFromTrade(trade, allowedSlippage, transactionDeadline)
+
+  const prevApprovalState = usePrevious(approvalState)
+
+  // mark when a user has submitted an approval, reset onTokenSelection for input field
+  useEffect(() => {
+    if (approvalState === ApprovalState.PENDING) {
+      setApprovalSubmitted(true)
+      // mod
+    } else if (prevApprovalState === ApprovalState.PENDING && approvalState === ApprovalState.NOT_APPROVED) {
+      // user canceled the approval tx, reset the UI
+      setApprovalSubmitted(false)
+    }
+  }, [approvalState, approvalSubmitted, prevApprovalState, setApprovalSubmitted])
+
+  const handleApprove = useCallback(async () => {
+    let approveRequired = false // mod
+    if (signatureState === UseERC20PermitState.NOT_SIGNED && gatherPermitSignature) {
+      try {
+        await gatherPermitSignature()
+      } catch (error) {
+        // try to approve if gatherPermitSignature failed for any reason other than the user rejecting it
+        if (!isRejectRequestProviderError(error)) {
+          approveRequired = true
+        }
+      }
+    } else {
+      approveRequired = true
+    }
+
+    if (approveRequired) {
+      const symbol = trade?.inputAmount?.currency.symbol
+      // const symbol = v2Trade?.inputAmount?.currency.symbol // TODO: double check
```

Comment:
Also remove?

Reply:
Fixed: https://github.com/cowprotocol/cowswap/pull/1049/commits/55332e1d0fbf0b1ab6871d98f8910f23542eb5c2
Also added TODO to refactor this code, I just separated it from the old code